src/librustdoc/html/markdown.rs

```diff
@@ -1057,7 +1057,14 @@ fn markdown_summary_with_limit(md: &str, length_limit: usize) -> (String, bool)
         *text_length += text.len();
     };
 
-    'outer: for event in Parser::new_ext(md, Options::ENABLE_STRIKETHROUGH) {
+    // NOTE: Make sure to update the same variable in `plain_text_summary`
+    //       if/when you update this one. They have to be duplicated because of a typesystem thing.
+    let mut broken_link_callback =
```

Comment:
What is 'a typesystem thing'? Can you paste the error?

Reply:
Sure. It involves a 'higher-ranked subtype' error 😨:

```
error[E0621]: explicit lifetime required in the type of `md`
    --> src/librustdoc/html/markdown.rs:1065:9
     |
1050 | fn markdown_summary_with_limit(md: &str, length_limit: usize) -> (String, bool) {
     |                                    ---- help: add explicit lifetime `'static` to the type of `md`: `&'static str`
...
1065 |         Parser::new_with_broken_link_callback(md, summary_opts(), Some(&mut summary_broken_link_callback))
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ lifetime `'static` required

error: higher-ranked subtype error
    --> src/librustdoc/html/markdown.rs:1065:72
     |
1065 |         Parser::new_with_broken_link_callback(md, summary_opts(), Some(&mut summary_broken_link_callback))
     |                                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0716]: temporary value dropped while borrowed
    --> src/librustdoc/html/markdown.rs:1065:77
     |
1065 |         Parser::new_with_broken_link_callback(md, summary_opts(), Some(&mut summary_broken_link_callback))
     |         --------------------------------------------------------------------^^^^^^^^^^^^^^^^^^^^^^^^^^^^--
     |         |                                                                   |                            |
     |         |                                                                   |                            temporary value is freed at the end of this statement
     |         |                                                                   creates a temporary which is freed while still in use
     |         argument requires that borrow lasts for `'static`
```

Reply:
For now you can fix the error like so:

```diff
diff --git a/src/librustdoc/html/markdown.rs b/src/librustdoc/html/markdown.rs
index 10f0befbc04..a61496be610 100644
--- a/src/librustdoc/html/markdown.rs
+++ b/src/librustdoc/html/markdown.rs
@@ -1035,6 +1035,10 @@ impl MarkdownSummaryLine<'_> {
     }
 }
 
+fn broken_link_callback<'a>(link: BrokenLink<'_>) -> Option<(CowStr<'a>, CowStr<'a>)> {
+  Some(("#".into(), link.reference.to_owned().into()))
+}
+
 /// Renders a subset of Markdown in the first paragraph of the provided Markdown.
 ///
 /// - *Italics*, **bold**, and `inline code` styles **are** rendered.
@@ -1057,11 +1061,6 @@ fn push(s: &mut String, text_length: &mut usize, text: &str) {
         *text_length += text.len();
     };
 
-    // NOTE: Make sure to update the same variable in `plain_text_summary`
-    //       if/when you update this one. They have to be duplicated because of a typesystem thing.
-    let mut broken_link_callback =
-        |broken_link: BrokenLink<'_>| Some(("#".into(), broken_link.reference.to_owned().into()));
-
     'outer: for event in
         Parser::new_with_broken_link_callback(md, summary_opts(), Some(&mut broken_link_callback))
     {
@@ -1142,11 +1141,6 @@ fn push(s: &mut String, text_length: &mut usize, text: &str) {
 
     let mut s = String::with_capacity(md.len() * 3 / 2);
 
-    // NOTE: Make sure to update the same variable in `markdown_summary_with_limit`
-    //       if/when you update this one. They have to be duplicated because of a typesystem thing.
-    let mut broken_link_callback =
-        |broken_link: BrokenLink<'_>| Some(("#".into(), broken_link.reference.to_owned().into()));
-
     for event in
         Parser::new_with_broken_link_callback(md, summary_opts(), Some(&mut broken_link_callback))
     {
```

The issue is that pulldown-cmark is tying the life of the callback to the life of the input, which I think is a bug. Working on a fix for them now, but it doesn't need to block this PR - in this case, the life of the input doesn't matter because you always call `to_owned()`. To avoid the unnecessary copy we'd need the bug to be fixed.

Reply:
This was more hairy than expected, I opened an issue instead. https://github.com/raphlinus/pulldown-cmark/issues/509